feat: add support to collect metrics using the OTEL library

zorocloud opened the pull request:
Adding support to the interceptor proxy to allow for the collection of metrics via a prometheus compatible endpoint or via a OTEL HTTP collector. Resolves #910 

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO)
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] Any necessary documentation is added, such as:
  - [`README.md`](/README.md)
  - [The `docs/` directory](./docs)
  - [The docs repo](https://github.com/kedacore/keda-docs)

JorTurFer reviewed (commented):
Looks really nice! Could you add some documentation about how to enable these metrics? I think that we could create a new section like `Operate` and include this information about how to enable prometheus/otel and which metrics are available.

do you think that adding an e2e test could be useful? In KEDA core we cover the metrics too using e2e tests, so we could just add an e2e test like them

zorocloud commented:
> Looks really nice! Could you add some documentation about how to enable these metrics? I think that we could create a new section like `Operate` and include this information about how to enable prometheus/otel and which metrics are available.
Thanks! Yeah, sure. Will draft some docs up and get the PR updated.

> do you think that adding an e2e test could be useful? In KEDA core we cover the metrics too using e2e tests, so we could just add an e2e test like them
That is a good shout. Will get a test for both cases added in and update the PR.

zorocloud commented:
I have added a brief overview of how to configure each exporter in a Operate.md doc as per your suggestion @JorTurFer and added in some e2e tests (using the core Keda set as a reference). Let me know your thoughts :)

zorocloud commented on config/interceptor/deployment.yaml:
```diff
@@ -50,11 +50,23 @@ spec:
           containerPort: 9090
         - name: proxy
           containerPort: 8080
+        - name: metrics
+          containerPort: 2223
```
It's the one they used in the OTEL example, just kept it the same tbh lol

JorTurFer commented on tests/checks/interceptor_prometheus_metrics/interceptor_prometheus_metrics_test.go:
```diff
@@ -0,0 +1,233 @@
+	assert.True(t, ok, "interceptor_request_count_total is available")
+
+	requestCount := getMetricsValue(val)
+	assert.Equal(t, float64(1), requestCount)
```
I'd assert this using greater or equal because the Prometheus metrics are persisted over the time, and 2nd and 3rd tries can fail because the metric returns 2,3, etc

zorocloud replied:
Yeah that is a good shout - when I had tested I was just tearing it down after every run so only ever had a count of 1 but that won't always be true

JorTurFer commented on tests/checks/interceptor_otel_metrics/interceptor_otel_metrics_test.go:
```diff
@@ -0,0 +1,242 @@
+      - "exec tail -f /dev/null"`
+)
+
+func TestMetricGeneration(t *testing.T) {
```
This e2e test has failed due to compiling issues

zorocloud replied:
Oops, left a small typo in there. Removed.

JorTurFer commented on tests/checks/interceptor_otel_metrics/interceptor_otel_metrics_test.go:
```diff
@@ -0,0 +1,242 @@
+	assert.True(t, ok, "interceptor_request_count_total is available")
+
+	requestCount := getMetricsValue(val)
+	assert.Equal(t, float64(1), requestCount)
```
I'd assert this using greater or equal because the Prometheus metrics are persisted over the time, and 2nd and 3rd tries can fail because the metric returns 2,3, etc

zorocloud commented:
Will take a look why the e2e tests failed when I get a sec

JorTurFer reviewed (approved):
LGTM!

JorTurFer commented:
There are some static code errors, could you take a look please? https://github.com/kedacore/http-add-on/actions/runs/8453796744/job/23280002430?pr=927

tomkerkhove commented on docs/operate.md:
```diff
@@ -0,0 +1,13 @@
+There are currently 2 supported methods for exposing metrics from the interceptor proxy service - via a Prometheus compatible metrics endpoint or by pushing metrics to a OTEL HTTP collector.
+
+### Configuring the Prometheus compatible metrics endpoint
+When configured the interceptor proxy can expose metrics on a Prometheus compatible endpoint. This endpoint can be enabled by setting the `KEDA_HTTP_OTEL_PROM_EXPORTER_ENABLED` environment variable to `true` on the interceptor deployment (`true` by default) and by setting `KEDA_HTTP_OTEL_PROM_EXPORTER_PORT` to an unused port for the endpoint to be made avaialble on (`2223` by default).
```
```suggestion
When configured, the interceptor proxy can expose metrics on a Prometheus compatible endpoint.

This endpoint can be enabled by setting the `KEDA_HTTP_OTEL_PROM_EXPORTER_ENABLED` environment variable to `true` on the interceptor deployment (`true` by default) and by setting `KEDA_HTTP_OTEL_PROM_EXPORTER_PORT` to an unused port for the endpoint to be made available on (`2223` by default).
```

zorocloud replied:
updated :)

tomkerkhove commented on docs/operate.md:
```diff
@@ -0,0 +1,13 @@
+When configured the interceptor proxy can expose metrics on a Prometheus compatible endpoint. This endpoint can be enabled by setting the `KEDA_HTTP_OTEL_PROM_EXPORTER_ENABLED` environment variable to `true` on the interceptor deployment (`true` by default) and by setting `KEDA_HTTP_OTEL_PROM_EXPORTER_PORT` to an unused port for the endpoint to be made avaialble on (`2223` by default).
+
+### Configuring the OTEL HTTP exporter
+When configured the interceptor proxy can export metrics to a OTEL HTTP collector. The OTEL exporter can be enabled by setting the `KEDA_HTTP_OTEL_HTTP_EXPORTER_ENABLED` environment variable to `true` on the interceptor deployment (`false` by default). When enabled the `KEDA_HTTP_OTEL_HTTP_COLLECTOR_ENDPOINT` environment variable must also be configured so the exporter knows what collector to sen ...
```
```suggestion
When configured, the interceptor proxy can export metrics to a OTEL HTTP collector.

The OTEL exporter can be enabled by setting the `KEDA_HTTP_OTEL_HTTP_EXPORTER_ENABLED` environment variable to `true` on the interceptor deployment (`false` by default). When enabled the `KEDA_HTTP_OTEL_HTTP_COLLECTOR_ENDPOINT` environment variable must also be configured so the exporter knows what collector to send the metrics to (e.g. opentelemetry-collector.open-telemetry-system:4318).

If the collector is exposed on an unsecured endpoint then you can set the `KEDA_HTTP_OTEL_HTTP_COLLECTOR_INSECURE` environment variable to `true` (`false` by default) which will disable client security on the exporter.

If you need to provide any headers such as authentication details in order to utilise your OTEL collector you can add them into the `KEDA_HTTP_OTEL_HTTP_HEADERS` environment variable. The frequency at which the metrics are exported can be configured by setting `KEDA_HTTP_OTEL_METRIC_EXPORT_INTERVAL` to the number of seconds you require between each export interval (`30` by default).
```

zorocloud replied:
Updated :)

tomkerkhove reviewed (commented):
Docs LGTM, nice work!

zorocloud commented:
Thanks for the feedback. Remaining suggestions should be addressed and the static checks fixed. Just noticed I need to fix some conflicts too. Will sort that when I next get a sec :+1:

zorocloud commented:
I have just rebased onto main and the unit/e2e tests look like they are passing fine so we should be good now 👍

JorTurFer commented:
We can ignore Synk, it has transient failures from time to time xD I have marked it as okey manually :)

JorTurFer commented:
There is other PR ongoing and once that's merged, I'll update your PR and merge it. Thanks for your AWESOME contribution ❤️

JorTurFer commented on tests/utils/setup_test.go:
```diff
@@ -14,6 +15,60 @@ import (
 
+var (
+	OtlpConfig = `mode: deployment
+config:
+  exporters:
```
```suggestion
image:
  repository: "otel/opentelemetry-collector-contrib"
config:
  exporters:
```

JorTurFer replied:
There is a potential breaking change that requires the change above, please commit it 🙏 
https://github.com/open-telemetry/opentelemetry-helm-charts/blob/main/charts/opentelemetry-collector/UPGRADING.md#0880-to-0890

zorocloud replied:
Done @JorTurFer :) 

Also, I've pushed a change to move the pending request count from a Gauge to a UpDownCounter (noticed some memory issues with the Gauge). Left it in a separate commit rather than squashing locally so it's easier for you to review.

JorTurFer commented:
There are merge conflicts to be solved, could you solve them? (or enabling me to update your PR xD)

zorocloud commented:
Should be sorted now @JorTurFer  :) 

Also, allowed maintainers to edit the PR, not sure why I didn't select this previously.

JorTurFer reviewed (approved):
LGTM! ❤️

JorTurFer commented:
I didn't noticed during the review that we aren't respected the OTEL spec ENV vars: https://opentelemetry.io/docs/specs/otel/protocol/exporter/

I've created an issue to track it, as I guess that you are using the metrics, I share it with you because we will have to change them for future versions: https://github.com/kedacore/http-add-on/issues/1031

zorocloud commented:
Ah sugar, I completely missed that too :facepalm:. I'm happy to get a PR raised toward the end of next week to fix this if someone isn't already working on it.

JorTurFer commented:
>I'm happy to get a PR raised toward the end of next week to fix this if someone isn't already working on it.

I'd be nice, I just pinged you to make you aware about a possible breaking change, but if you are willing to update the code, it's awesome :)
